### ISSUE TITLE
Build on Windows not working - Deltachat.dll missing

### ISSUE THREAD

**Issue by cracker0dks** [deltachat/deltachat-desktop#1726]
Followed this guid: https://github.com/deltachat/deltachat-desktop/blob/master/docs/DEVELOPMENT.md
and run: `npm run dev`

I get this error:
```
E:\Code\deltachat-desktop>npm run dev

> deltachat-desktop@1.4.0 dev E:\Code\deltachat-desktop
> npm run build && npm run start --


> deltachat-desktop@1.4.0 build E:\Code\deltachat-desktop
> npm run build:shared && npm run build:backend && npm run build:frontend


> deltachat-desktop@1.4.0 build:shared E:\Code\deltachat-desktop
> npm run build:shared:translations && npm run build:shared:version


> deltachat-desktop@1.4.0 build:shared:translations E:\Code\deltachat-desktop
> node ./bin/build-shared-convert-translations.js ./_locales

+ Converting translations from xml to json...
+ converted 36 files in 395 ms


> deltachat-desktop@1.4.0 build:shared:version E:\Code\deltachat-desktop
> node ./bin/build-shared-version-info.js


> deltachat-desktop@1.4.0 build:backend E:\Code\deltachat-desktop
> tsc --build src/main

src/shared/shared-types.d.ts:61:36 - error TS2307: Cannot find module 'deltachat-node'.

61 import DeltaChat, { Contact } from 'deltachat-node'
                                      ~~~~~~~~~~~~~~~~

src/shared/shared-types.d.ts:91:31 - error TS2307: Cannot find module 'deltachat-node'.

91 import { Chat, Message } from 'deltachat-node'
                                 ~~~~~~~~~~~~~~~~


Found 2 errors.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! deltachat-desktop@1.4.0 build:backend: `tsc --build src/main`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the deltachat-desktop@1.4.0 build:backend script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Cracker\AppData\Roaming\npm-cache\_logs\2020-06-16T16_11_40_748Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! deltachat-desktop@1.4.0 build: `npm run build:shared && npm run build:backend && npm run build:frontend`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the deltachat-desktop@1.4.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Cracker\AppData\Roaming\npm-cache\_logs\2020-06-16T16_11_40_794Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! deltachat-desktop@1.4.0 dev: `npm run build && npm run start --`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the deltachat-desktop@1.4.0 dev script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Cracker\AppData\Roaming\npm-cache\_logs\2020-06-16T16_11_40_837Z-debug.log
```

I also run `npm install` and tried to use other node versions like v13.14.0, v10.21.0, v8.16.0
through nvm without luck.

please help, thanks :)

**Comment by Jikstra** [comment]
you first need to run `npm install` to download/build the dependencies. What is the output of `npm install`?

**Comment by cracker0dks** [comment]
get this:
```
E:\Code\deltachat-desktop>npm install

> deltachat-node@1.35.0 install E:\Code\deltachat-desktop\node_modules\deltachat-node
> node-gyp-build "npm run build:core" && npm run build:bindings:c:postinstall && npm run build:bindings:ts 


> deltachat-node@1.35.0 build:bindings:c:postinstall E:\Code\deltachat-desktop\node_modules\deltachat-node
> node scripts/postinstall.js

E:\Code\deltachat-desktop\node_modules\deltachat-node\scripts\postinstall.js:42
  if (err) throw err
           ^

Error: ENOENT: no such file or directory, stat 'E:\Code\deltachat-desktop\node_modules\deltachat-node\deltachat-core-rust\target\release\deltachat.dll'
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! deltachat-node@1.35.0 build:bindings:c:postinstall: `node scripts/postinstall.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the deltachat-node@1.35.0 build:bindings:c:postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Cracker\AppData\Roaming\npm-cache\_logs\2020-06-16T17_00_51_228Z-debug.log
npm WARN The package tempy is included as both a dev and production dependency.
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.9 (node_modules\fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.9: wanted {"os":"darwin","arch":"any"} (current: {"os":"win32","arch":"x64"})

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! deltachat-node@1.35.0 install: `node-gyp-build "npm run build:core" && npm run build:bindings:c:postinstall && npm run build:bindings:ts `
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the deltachat-node@1.35.0 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Cracker\AppData\Roaming\npm-cache\_logs\2020-06-16T17_00_52_397Z-debug.log
```

**Comment by Jikstra** [comment]
This looks like a deltachat-node release, i was a bit aggressive on ignoring files for the last release because npm didnt allow us uploading the release due to the size. One way around it would be to use `npm install --build-from-source`.

**Comment by cracker0dks** [comment]
ok now I'm stuck here:
```
   Finished release [optimized] target(s) in 9m 10s

> deltachat-node@1.35.0 build:core:constants E:\Code\deltachat-desktop\node_modules\deltachat-node
> node scripts/generate-constants.js

Generating constants...

E:\Code\deltachat-desktop\node_modules\deltachat-node>if not defined npm_config_node_gyp (node "C:\Users\Cracker\AppData\Roaming\nvm\v8.16.0\node_modules\npm\node_modules\npm-lifecycle\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "C:\Users\Cracker\AppData\Roaming\nvm\v8.16.0\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js" rebuild )       
gyp ERR! configure error
gyp ERR! stack Error: Command failed: C:\Users\Cracker\AppData\Local\Programs\Python\Python38-32\python.EXE -c import sys; print "%s.%s.%s" % sys.version_info[:3];
gyp ERR! stack   File "<string>", line 1
gyp ERR! stack     import sys; print "%s.%s.%s" % sys.version_info[:3];
gyp ERR! stack                       ^
gyp ERR! stack SyntaxError: invalid syntax
gyp ERR! stack
gyp ERR! stack     at ChildProcess.exithandler (child_process.js:281:12)
gyp ERR! stack     at emitTwo (events.js:126:13)
gyp ERR! stack     at ChildProcess.emit (events.js:214:7)
gyp ERR! stack     at maybeClose (internal/child_process.js:915:16)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:209:5)
gyp ERR! System Windows_NT 10.0.18363
gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Users\\Cracker\\AppData\\Roaming\\nvm\\v8.16.0\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd E:\Code\deltachat-desktop\node_modules\deltachat-node
gyp ERR! node -v v8.16.0
gyp ERR! node-gyp -v v3.8.0
gyp ERR! not ok
npm WARN The package tempy is included as both a dev and production dependency.
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.9 (node_modules\fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.9: wanted {"os":"darwin","arch":"any"} (current: {"os":"win32","arch":"x64"})

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! deltachat-node@1.35.0 install: `node-gyp-build "npm run build:core" && npm run build:bindings:c:postinstall && npm run build:bindings:ts `
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the deltachat-node@1.35.0 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Cracker\AppData\Roaming\npm-cache\_logs\2020-06-16T17_22_48_693Z-debug.log
```
do I need python 2.7 ?

**Comment by Jikstra** [comment]
Yes and rust and probably some more dependencies :/

**Comment by Jikstra** [comment]
Probably checkout the v1.3.3 tag, this version should build on windows.

**Comment by Simon-Laux** [comment]
Try https://www.npmjs.com/package/windows-build-tools that works most of the time for installing the build requirements of native modules

**Comment by Jikstra** [comment]
Can reproduce this on ci: https://github.com/deltachat/deltachat-desktop/pull/1731

**Comment by cracker0dks** [comment]
Thanks for your help, installed "windows-build-tools" and checked out v1.3.3, than `npm install --build-from-source` had no more errors.

Now with `npm run dev` I was stuck here: 
```
E:\Code\deltachat-desktop>npm run dev

> deltachat-desktop@1.3.3 dev E:\Code\deltachat-desktop
> npm run build && npm run start --


> deltachat-desktop@1.3.3 build E:\Code\deltachat-desktop
> npm run build-main && npm run builder


> deltachat-desktop@1.3.3 build-main E:\Code\deltachat-desktop
> tsc --build src/main


> deltachat-desktop@1.3.3 builder E:\Code\deltachat-desktop
> node ./bin/build/builder.js

{ Error: spawn npx ENOENT
    at Process.ChildProcess._handle.onexit (internal/child_process.js:190:19)
    at onErrorNT (internal/child_process.js:362:16)
    at _combinedTickCallback (internal/process/next_tick.js:139:11)
    at process._tickDomainCallback (internal/process/next_tick.js:219:9)
    at Function.Module.runMain (module.js:696:11)
    at startup (bootstrap_node.js:204:16)
    at bootstrap_node.js:625:3
  errno: 'ENOENT',
  code: 'ENOENT',
  syscall: 'spawn npx',
  path: 'npx',
  spawnargs:
   [ 'node-sass',
     'scss/manifest.scss',
     'html-dist/main.css',
     '--output-style',
     'compressed',
     '--source-map',
     'true' ] }
{ Error: spawn npx ENOENT
    at Process.ChildProcess._handle.onexit (internal/child_process.js:190:19)
    at onErrorNT (internal/child_process.js:362:16)
    at _combinedTickCallback (internal/process/next_tick.js:139:11)
    at process._tickDomainCallback (internal/process/next_tick.js:219:9)
    at Function.Module.runMain (module.js:696:11)
    at startup (bootstrap_node.js:204:16)
    at bootstrap_node.js:625:3
  errno: 'ENOENT',
  code: 'ENOENT',
  syscall: 'spawn npx',
  path: 'npx',
  spawnargs: [ 'tsc', '-b', 'src/renderer', '--pretty' ] }
DONE: copy files from static folder
Converting translations from xml to json... (converted 36 files in 406 ms)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! deltachat-desktop@1.3.3 builder: `node ./bin/build/builder.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the deltachat-desktop@1.3.3 builder script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Cracker\AppData\Roaming\npm-cache\_logs\2020-06-18T14_14_03_214Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! deltachat-desktop@1.3.3 build: `npm run build-main && npm run builder`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the deltachat-desktop@1.3.3 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Cracker\AppData\Roaming\npm-cache\_logs\2020-06-18T14_14_03_257Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! deltachat-desktop@1.3.3 dev: `npm run build && npm run start --`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the deltachat-desktop@1.3.3 dev script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Cracker\AppData\Roaming\npm-cache\_logs\2020-06-18T14_14_03_294Z-debug.log
```
But running anything on master was a success \o.o/ 👍 thanks again!

**Comment by Jikstra** [comment]
@Simon-Laux any ideas?

@cracker0dks none of us is developing on windows, and windows often needs some extra love here and there. I think @Simon-Laux fixed something in the build scripts, probably you need to backport that.

**Comment by Jikstra** [comment]
Keeping this one open as it's a problem for the 1.4 series

**Comment by Simon-Laux** [comment]
@cracker0dks please use the current master and link to a local clone of deltachat-node
search deltachat node in the package.json and replace the version with the (relative) to your locally checked out deltachat node

@Jikstra the postinstall script for windows is broken. I suggest that we rethink how we do the prebuilds.

**Comment by Jikstra** [comment]
@cracker0dks should be fixed with the deltachat-node v1.39.0 release. Please reopen if it still happens on latest master :)
